File: go/tests/games_runtime_test.go
// games_runtime_test.go
//
// Verifies the CircleAI.Games port (games_runtime.go): scene-graph add/remove/
// snapshot, input-map fan-out + unsubscribe, timer game-loop tick delivery, and
// the null implementations.

package circleai_test

import (
	"context"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestGames_SceneGraph(t *testing.T) {
	g := circleai.NewInMemorySceneGraph()
	ctx := context.Background()
	if g.BackendId() != "in-memory" {
		t.Fatalf("backend id = %q", g.BackendId())
	}
	if err := g.Add(ctx, circleai.SceneNode{NodeId: "n1", Kind: "sprite", X: 1, Y: 2, Z: 3}); err != nil {
		t.Fatalf("add: %v", err)
	}
	if err := g.Add(ctx, circleai.SceneNode{NodeId: "n2", Kind: "light"}); err != nil {
		t.Fatalf("add: %v", err)
	}
	if err := g.Add(ctx, circleai.SceneNode{NodeId: "  "}); err == nil {
		t.Fatalf("blank NodeId must error")
	}
	snap, _ := g.Snapshot(ctx)
	if len(snap) != 2 {
		t.Fatalf("snapshot size = %d, want 2", len(snap))
	}
	if err := g.Remove(ctx, "n1"); err != nil {
		t.Fatalf("remove: %v", err)
	}
	snap, _ = g.Snapshot(ctx)
	if len(snap) != 1 || snap[0].NodeId != "n2" {
		t.Fatalf("post-remove snapshot failed: %+v", snap)
	}
}

func TestGames_InputMapFanOut(t *testing.T) {
	m := circleai.NewInMemoryInputMap()
	got := make(chan circleai.InputEvent, 8)
	// Subscribe synchronously before raising.
	unsub := m.Subscribe(func(ev circleai.InputEvent) { got <- ev })
	m.Raise(circleai.InputEvent{Action: "jump"})
	select {
	case ev := <-got:
		if ev.Action != "jump" {
			t.Fatalf("event action = %q", ev.Action)
		}
	case <-time.After(2 * time.Second):
		t.Fatal("timed out waiting for input event")
	}
	// After unsubscribe, no more events are delivered.
	unsub()
	m.Raise(circleai.InputEvent{Action: "fire"})
	select {
	case ev := <-got:
		t.Fatalf("unexpected event after unsubscribe: %+v", ev)
	case <-time.After(150 * time.Millisecond):
	}
}

func TestGames_TimerLoop(t *testing.T) {
	l := circleai.NewTimerGameLoop()
	ticks := make(chan circleai.GameTick, 16)
	// Subscribe synchronously before starting the loop.
	l.Subscribe(func(tk circleai.GameTick) { ticks <- tk })
	ctx := context.Background()
	if err := l.StartAsync(ctx, 100); err != nil {
		t.Fatalf("start: %v", err)
	}
	if err := l.StartAsync(ctx, 100); err == nil {
		t.Fatalf("double start must error")
	}
	select {
	case tk := <-ticks:
		if tk.Frame < 1 {
			t.Fatalf("first frame = %d, want >= 1", tk.Frame)
		}
	case <-time.After(2 * time.Second):
		t.Fatal("timed out waiting for a tick")
	}
	if err := l.StopAsync(ctx); err != nil {
		t.Fatalf("stop: %v", err)
	}
	_ = l.Close()
}

func TestGames_NullImpls(t *testing.T) {
	ctx := context.Background()
	var loop circleai.GameLoop = circleai.NullGameLoop{}
	if loop.BackendId() != "null" {
		t.Fatalf("null loop backend id = %q", loop.BackendId())
	}
	if err := loop.StartAsync(ctx, 60); err != nil {
		t.Fatalf("null start: %v", err)
	}
	unsub := loop.Subscribe(func(circleai.GameTick) {})
	unsub()

	var im circleai.InputMap = circleai.NullInputMap{}
	im.Subscribe(func(circleai.InputEvent) {})()

	var sg circleai.SceneGraph = circleai.NullSceneGraph{}
	if err := sg.Add(ctx, circleai.SceneNode{NodeId: "x"}); err != nil {
		t.Fatalf("null scene add: %v", err)
	}
	if snap, _ := sg.Snapshot(ctx); len(snap) != 0 {
		t.Fatalf("null scene snapshot must be empty: %+v", snap)
	}
}

File: typescript/tests/games_runtime.test.ts
// games_runtime.test.ts
// Verifies the CircleAI.Games port: timer game loop start/stop + fan-out +
// unsubscribe, in-memory input map, in-memory scene graph, and null impls.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  TimerGameLoop,
  InMemoryInputMap,
  InMemorySceneGraph,
  NullGameLoop,
  NullInputMap,
  NullSceneGraph,
  sceneNode,
  inputEvent,
  type GameTick,
} from "../src/games/index";

const sleep = (ms: number): Promise<void> => new Promise((r) => setTimeout(r, ms));

describe("TimerGameLoop", () => {
  it("fans ticks out to subscribers with monotonically increasing frames", async () => {
    const loop = new TimerGameLoop();
    assert.equal(loop.backendId, "timer");
    const frames: number[] = [];
    loop.subscribe((t: GameTick) => {
      frames.push(t.frame);
    });
    await loop.startAsync(200); // ~5ms period
    await sleep(60);
    await loop.stopAsync();
    assert.ok(frames.length >= 2, `expected several ticks, got ${frames.length}`);
    assert.equal(frames[0], 1);
    for (let i = 1; i < frames.length; i++) assert.equal(frames[i], frames[i - 1] + 1);
    await loop.disposeAsync();
  });

  it("rejects a second start and non-positive fps", async () => {
    const loop = new TimerGameLoop();
    await loop.startAsync(100);
    assert.throws(() => void loop.startAsync(100), /already started/);
    await loop.stopAsync();
    assert.throws(() => void loop.startAsync(0), /targetFps/);
  });

  it("stops delivering after unsubscribe", async () => {
    const loop = new TimerGameLoop();
    let count = 0;
    const sub = loop.subscribe(() => {
      count++;
    });
    await loop.startAsync(200);
    await sleep(40);
    sub.dispose();
    sub.dispose(); // idempotent
    const seen = count;
    await sleep(40);
    await loop.stopAsync();
    assert.equal(count, seen, "no ticks should arrive after dispose");
  });
});

describe("InMemoryInputMap", () => {
  it("raises events to all subscribers and stops after dispose", () => {
    const map = new InMemoryInputMap();
    assert.equal(map.backendId, "in-memory");
    const seen: string[] = [];
    const sub = map.subscribe((ev) => {
      seen.push(ev.action);
    });
    map.raise(inputEvent("jump"));
    map.raise(inputEvent("crouch", new Map([["key", "ctrl"]])));
    sub.dispose();
    map.raise(inputEvent("shoot"));
    assert.deepEqual(seen, ["jump", "crouch"]);
  });
});

describe("InMemorySceneGraph", () => {
  it("adds, snapshots, and removes nodes", async () => {
    const g = new InMemorySceneGraph();
    assert.equal(g.backendId, "in-memory");
    await g.addAsync(sceneNode("n1", "sprite", 1, 2, 3));
    await g.addAsync(sceneNode("n2", "light", 4, 5, 6));
    let snap = await g.snapshotAsync();
    assert.deepEqual(snap.map((n) => n.nodeId).sort(), ["n1", "n2"]);
    await g.removeAsync("n1");
    snap = await g.snapshotAsync();
    assert.deepEqual(snap.map((n) => n.nodeId), ["n2"]);
  });

  it("rejects blank node ids", async () => {
    const g = new InMemorySceneGraph();
    await assert.rejects(() => g.addAsync(sceneNode("  ", "x", 0, 0, 0)), /NodeId required/);
    await assert.rejects(() => g.removeAsync(""), /nodeId required/);
  });
});

describe("Null implementations", () => {
  it("are inert and report backend 'null'", async () => {
    const loop = new NullGameLoop();
    assert.equal(loop.backendId, "null");
    await loop.startAsync();
    await loop.stopAsync();
    const d = loop.subscribe(() => {});
    d.dispose();
    await loop.disposeAsync();

    assert.equal(NullInputMap.instance.backendId, "null");
    NullInputMap.instance.subscribe(() => {}).dispose();

    assert.equal(NullSceneGraph.instance.backendId, "null");
    await NullSceneGraph.instance.addAsync(sceneNode("n1", "x", 0, 0, 0));
    await NullSceneGraph.instance.removeAsync("n1");
    assert.deepEqual(await NullSceneGraph.instance.snapshotAsync(), []);
  });
});
